sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AsynchronousMetricStorage.java

```diff
@@ -168,25 +219,99 @@ public MetricData collect(
       InstrumentationScopeInfo instrumentationScopeInfo,
       long startEpochNanos,
       long epochNanos) {
-    Map<Attributes, T> result;
+    if (memoryMode == REUSABLE_DATA) {
+      // Collect can not run concurrently for same reader, hence we safely assume
+      // the previous collect result has been used and done with
+      reusableResultList.forEach(v -> reusablePointsPool.returnObject(v));
+      reusableResultList.clear();
+    }
+
+    Collection<T> result;
     if (aggregationTemporality == AggregationTemporality.DELTA) {
       Map<Attributes, T> points = this.points;
       Map<Attributes, T> lastPoints = this.lastPoints;
-      lastPoints.entrySet().removeIf(entry -> !points.containsKey(entry.getKey()));
+
+      Collection<T> deltaPoints;
+      switch (memoryMode) {
+        case REUSABLE_DATA:
+          deltaPoints = reusableResultList;
+          break;
+        case IMMUTABLE_DATA:
+          deltaPoints = new ArrayList<>();
+          break;
+        default:
+          throw new IllegalStateException("Unsupported memory mode: " + memoryMode);
+      }
+
       points.forEach(
-          (k, v) -> lastPoints.compute(k, (k2, v2) -> v2 == null ? v : aggregator.diff(v2, v)));
-      result = lastPoints;
+          (k, v) -> {
+            T lastPoint = lastPoints.get(k);
+
+            T deltaPoint;
+            if (lastPoint == null) {
+              switch (memoryMode) {
+                case REUSABLE_DATA:
+                  deltaPoint = reusablePointsPool.borrowObject();
+                  aggregator.copyPoint(v, deltaPoint);
+                  break;
+                case IMMUTABLE_DATA:
+                  deltaPoint = v;
+                  break;
+                default:
+                  throw new IllegalStateException("Unsupported memory mode: " + memoryMode);
+              }
+            } else {
+              switch (memoryMode) {
+                case REUSABLE_DATA:
+                  aggregator.diffInPlace(lastPoint, v);
```

Comment:
Instead of adding a new `diffInPlace` method, can we make the implementation of the existing `diff` implementation smarter by detecting whether its an mutable or immutable copy, and either updating in place or creating a new immutable record?

Reply:
Yes, it's possible, and diff would return the mutable one.
The reason I've decided to do it since I wanted to make the code more obvious / explicit.
When you read this section:
```
              if (memoryMode == REUSABLE_DATA) {
                aggregator.diffInPlace(lastPoint, v);
                deltaPoint = lastPoint;

                // Remaining last points are returned to reusablePointsPool, but
                // this reusable point is still used, so don't return it to pool yet
                lastPoints.remove(k);
              } else {
                deltaPoint = aggregator.diff(lastPoint, v);
              }
```
You say: "so it's reusable data" and you see `aggregator.diffInPlace(lastPoint, v);` and you immediately understand the result will be placed inside v (well that's what aimed for).

If would have something like:
```
              if (memoryMode == REUSABLE_DATA) {
                aggregator.diff(lastPoint, v);
                deltaPoint = lastPoint;

                // Remaining last points are returned to reusablePointsPool, but
                // this reusable point is still used, so don't return it to pool yet
                lastPoints.remove(k);
              } else {
                deltaPoint = aggregator.diff(lastPoint, v);
              }
```

Would it be obvious that `aggregator.diff(lastPoint, v)` will save the diff result into `v`?
Maybe it should be `aggregator.diffIntoReusable(lastPoint, v)`?

I just wanted to avoid people have to peer into the "diff" implementation, keep in mind, go back and read.
I'm not sure. What is your opinion?